### PR TITLE
Agroal DB pool test coverage

### PR DESCRIPTION
**Pull request by pjgg**
- Add some test coverage in order to double-check Agroal DB pool management.

### CODE REVIEW

File: 014-quarkus-panache-with-transactions-xa/src/main/resources/application.properties
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/main/resources/application.properties:10]: Nack: Setting `max-size` to 1 was done on purpose to ensure no more than 1 connection is created. Indeed, thanks to this setting, we have been able to spot issues on Quarkus upstream.

**Comment by jsmrcka** [comment]
Got the same fail locally as in GH checks:
```
org.opentest4j.AssertionFailedError: Expected 1 active connections ==> expected: <1> but was: <3>
	at io.quarkus.qe.dbpool.AgroalPoolTest.agroalCheckIdleTimeout(AgroalPoolTest.java:82)
```

**Comment by jsmrcka** [comment]
I'm missing a mention in README about the new test. Some kind of description of does the test verify.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (jsmrcka) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:88]: I find this misleading. `completed.getCount()` returns number of unfinished events, which is by itself a little bit confusing naming. And here, `events - completed.getCount()` returns number of finished events, opposite to what the message states.

This is also the case in `agroalCheckBorderConditionBetweenIdleAndGetConnection`

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (jsmrcka) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:97]: Consider adding a prefix to set it apart from identical log entry in `agroalConcurrentLoadTest`.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (jsmrcka) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:103]: Consider adding a prefix to set it apart from identical log entry in `agroalConcurrentLoadTest`.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java]: good catch!

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (jsmrcka) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:42]: The class name already declares that it is about Agroal. So I would drop the `agroal` prefix in test method names.
Also, it would be nice to name all the test methods uniformly, either `testSomething`, `checkSomething` or just `something`. Now it is a mixture.
WDYT?

**Review by jsmrcka — CHANGES_REQUESTED** [review]

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (jsmrcka) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:116]: I think the formatting check is failing on this line being too long.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (jsmrcka) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:142]: I think the formatting check is failing on this line being too long.

**Review by jsmrcka — APPROVED** [review]
LGTM. Thanks :).

File: 014-quarkus-panache-with-transactions-xa/README.md
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/README.md:20]:
```suggestion
- AgroalPoolTest will cover how the db pool is manage in terms of IDLE-timeout, max connections and concurrency. 
```

Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/README.md:20]:
```suggestion
- AgroalPoolTest will cover how the db pool is managed in terms of IDLE-timeout, max connections and concurrency. 
```

File: 014-quarkus-panache-with-transactions-xa/src/main/resources/application.properties
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/main/resources/application.properties:21]: Two emtpy lines here.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/containers/MysqlTestProfile.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/containers/MysqlTestProfile.java:8]: QuarkusTestProfile is not compatible when running the test on Native, so we're missing this coverage on Native. Do you think this coverage is not important for Native, so we can live without it?

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:41]: Is the test method order a requirement to run this test? Having tests that depend on each other is most of the times a bad practice.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:70]: Why 3 seconds? If this value depends on a property, I would rather using this property instead of a fixed value.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:77]: I think `final` is redundant.
Also, why 2500 events? It seems to me like too much.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:78]: I think `final` is redundant.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:103]: I think `final` is redundant.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:112]: I don't understand very well this code.
Are you running two concurrent threads and each thread is making calls to the application every 3 seconds? Then, we verify that the queuedEvents is 0... 
Why do we need 2 threads? If one of them fails, the another thread can call the application twice and then the test will still satisfy the condition.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:124]: This test depends on the execution of other tests which is generally a bad practice. Can we assert this condition as part of the other code instead?

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:132]: Same comment as in the method borderConditionBetweenIdleAndGetConnectionTest

**Review by Sgitario — CHANGES_REQUESTED** [review]
Apart from the comments, the most important topics are:
- We're missing the Native coverage and from my point of view this would be necessary as well.
- Why are you using the `@RepeatedTest` annotation? Repeating the test too many times seems too much for me.
- Tests depend on the method order to work fine and generally speaking this is a bad practice.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/containers/MysqlTestProfile.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/containers/MysqlTestProfile.java:8]: is not intended to be a native test.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java]: Not really is pending to a Zulip resolution. I have a question related to `quarkus.datasource.jdbc.idle-removal-interval`. But I agree we can disable this test and remove the order stuff until having a reply.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:111]: is not every 3 sec. We are making a query 3 ms ahead of the IDLE-timeout. 

`.every(Duration.ofSeconds(getIdle() + 3))`

We are checking border condition errors.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:123]: All tests are an isolated tests.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:111]: I still don't understand this method. As I can read: `.every(Duration.ofSeconds(1 + 3))`, It's traslated to every 4 seconds. If it's for some reason, it's indeed 3 or 4 ms, can you make it more readable please?

**Review by Sgitario — CHANGES_REQUESTED** [review]
Some comments are marked as resolved, but they're not.
And this comment https://github.com/quarkus-qe/beefy-scenarios/pull/203/files#r632530612 is still unclear to me.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:111]: Ok I have changed a little bit the signature in order to be more readable:

`.every(Duration.ofSeconds(getIdleMs() + 3))`

- 2 concurrent threads and then making a request every X time: 

concurrent queries every X secs N times (Repeat annotation) is exactly the aim of this test. This is why is called `connectionConcurrencyTest`, or in other words, check that two (or more) concurrent queries completed successfully. In case of any issue, ` queuedEvents.countDown()` is not going to be invoked and the test is going to fail 

```
assertEquals(0, queuedEvents.getCount(),
                String.format("agroalConnectionConcurrency: Missing %d events.", queuedEvents.getCount()));
```

 - If what we want is to stress the db pool, what about running this test using more than 2 threads (which I think it simulates the users):
 
  we could increase the number of concurrent queries (I did , is more expensive in terms of CPU and time, I have increased from 2 to 20) but, the `stress` feature is cover by the `RepeatedTest` annotation (is a combination of both). So, 2 concurrent queries repeated 500 times, means 1000 consecutive successful queries. 

With the current concurrent configuration, attending to your request, we are going to make 10k consecutive successful queries.  Is going to be more expensive(10 times), I thought that 1000 queries were enough to probe that we handler concurrency as expected ( some issues have risen with even less iteration and concurrency), from a agroal datasource point of view, but is also ok)

**Comment by pjgg** [comment]
> Some comments are marked as resolved, but they're not.
> And this comment https://github.com/quarkus-qe/beefy-scenarios/pull/203/files#r632530612 is still unclear to me.

Which comment si is not clear? 

this one? This test depends on the execution of other tests which is generally a bad practice. Can we assert this condition as part of the other code instead?

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:111]: I see you have changed the every method to:
`.every(Duration.ofSeconds(getIdleMs() + 3))`

But it's still odd to me as you have milliseconds and use it as seconds.
Let me be clearer, this is what you are doing:

```
int concurrentRequestAmount = 2; // To mimic users
int idle = 1; // 1 seconds
long idleMs = Duration.ofSeconds(idle).toMillis(); // 1000 milliseconds
CountDownLatch queuedEvents = new CountDownLatch(concurrentRequestAmount); // Wait for 2 requests
Multi.createFrom().range(0, concurrentRequestAmount).subscribe()
        .with(n -> Multi.createFrom().ticks()
                .every(Duration.ofSeconds(idleMs + 3)) // 1000 + 3 as seconds 
                .onOverflow().drop()
                .subscribe().with(i -> {
                    System.out.println("Tick: " + i);
                    queuedEvents.countDown();
                }));
assertTrue(queuedEvents.await(5, TimeUnit.SECONDS));     
```

This is working for you, because you're waiting for 2 requests done by 2 concurrent requests, but the `Multi.createFrom().ticks().every(Duration.ofSeconds(idleMs + 3))` is doing a request every 1003 seconds now (before it was 4 seconds).

I think the right code is to change it to `Multi.createFrom().ticks().every(Duration.ofMillis(idleMs + 3))`.

However, as you are verifying that the requests of the concurrent users worked fine, the current solution would be the same as doing:

```
final int concurrentRequestAmount = 20;
CountDownLatch queuedEvents = new CountDownLatch(concurrentRequestAmount);
Multi.createFrom().range(0, concurrentRequestAmount).subscribe().with(n -> {
            makeApplicationQuery();
            queuedEvents.countDown();
        });
```

But guessing that what you exactly want to stress the border condition, then another idea would be to either assert there are no failures after some amount of time or verify the requests after some ticks (`new CountDownLatch(concurrentRequestAmount * NUM_OF_ITERATIONS or ticks);`

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (Sgitario) [inline 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:71]: Is this working? You are parsing 1000 milliseconds + 1, then converting it to seconds, which is 1001 seconds and then to milliseconds 1001000. So, you're sleeping the test for 1001000 milliseconds.

**Review by Sgitario — CHANGES_REQUESTED** [review]

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java]: Was leftover as a result of a side effect from your previous comments.

File: 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java
Comment (pjgg) [reply 014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java:111]: Yes in this test `borderConditionBetweenIdleAndGetConnectionTest` the key concept is the IDLE-timeout border condition.

**Review by Sgitario — APPROVED** [review]

**Comment by Sgitario** [comment]
Waiting for CI before merging.

**Review by Sgitario — CHANGES_REQUESTED** [review]
The CI action failed for Native. I'm aware that the changes do no affect the Native build, but I think it spotted an issue as the AgroalPoolTest test was stucked. From logs:

```
2021-05-17T09:09:46.9832375Z 2021-05-17 09:09:42,758 WARN  [io.agr.pool] (executor-thread-39) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
2021-05-17T09:09:46.9833909Z 2021-05-17 09:09:42,759 INFO  [io.qua.qe.dbp.AgroalPoolTest] (executor-thread-103) 2021-05-17 09:09:42
2021-05-17T09:09:46.9835516Z 2021-05-17 09:09:42,759 WARN  [io.agr.pool] (executor-thread-103) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
2021-05-17T09:09:46.9838630Z 2021-05-17 09:09:42,759 INFO  [io.qua.qe.dbp.AgroalPoolTest] (executor-thread-112) 2021-05-17 09:09:42
2021-05-17T09:09:46.9843561Z 2021-05-17 09:09:42,759 WARN  [io.agr.pool] (executor-thread-112) Datasource '<default>': JDBC resources leaked: 0 ResultSet(s) and 1 Statement(s)
2021-05-17T09:09:46.9847045Z 2021-05-17 09:09:42,759 INFO  [io.qua.qe.dbp.AgroalPoolTest] (executor-thread-112) 2021-05-17 09:09:42
```

**Review by Sgitario — APPROVED** [review]

**Pull request merged**
